rsciio/jobin_yvon/api.py

```diff
@@ -0,0 +1,819 @@
+        self.axes = dict()
+        self._has_nav1 = False
+        self._has_nav2 = False
+        for child in self._axis_root:
+            if self._get_id(child) == "0x0":
+                self._set_signal_type(child)
+            if self._get_id(child) == "0x1":
+                self._set_signal_axis(child)
+            if self._get_id(child) == "0x2":
+                self._has_nav1, self._nav1_size = self._set_nav_axis(child, "nav1_dict")
+            if self._get_id(child) == "0x3":
+                self._has_nav2, self._nav2_size = self._set_nav_axis(child, "nav2_dict")
+
+        self._sort_nav_axes()
+
+    def get_data(self):
+        """Extract data from file."""
+        data_raw = self._lsx_matrix.findall("LSX_Row")
+        ## lexicographical ordering -> 3x3 map -> 9 rows
+        num_rows = len(data_raw)
+        if num_rows == 0:
+            _logger.critical("No data found.")  # pragma: no cover
+        elif num_rows == 1:
+            ## Spectrum
+            self.data = np.fromstring(data_raw[0].text.strip(), sep=" ")
+            if self._reverse_signal:
+                self.data = self.data[::-1]
+        else:
+            ## linescan or map
+            num_cols = self._get_size(data_raw[0])
+            self.data = np.empty((num_rows, num_cols))
+            for i, row in enumerate(data_raw):
+                row_array = np.fromstring(row.text.strip(), sep=" ")
+                if self._reverse_signal:
+                    row_array = row_array[::-1]
+                self.data[i, :] = row_array
+            ## reshape the array (lexicographic -> cartesian)
+            ## reshape depends on available axes
+            if self._has_nav2:
+                if self._has_nav1:
+                    self.data = np.reshape(
+                        self.data, (self._nav2_size, self._nav1_size, num_cols)
+                    )
+                else:
+                    self.data = np.reshape(
+                        self.data, (self._nav2_size, num_cols)
+                    )  # pragma: no cover
+            elif self._has_nav1 and not self._has_nav2:
+                self.data = np.reshape(self.data, (self._nav1_size, num_cols))
+
+    ## account for missing keys in original metadata (try/except)
+    def map_metadata(self):
+        """Maps original_metadata to metadata dictionary."""
+        self.metadata = dict()
+        if "General" not in self.metadata:
+            self.metadata["General"] = {}
+        if "Signal" not in self.metadata:
+            self.metadata["Signal"] = {}
+        if "Sample" not in self.metadata:
+            self.metadata["Sample"] = {}
+        if "Acquisition_instrument" not in self.metadata:
+            self.metadata["Acquisition_instrument"] = {
+                "Laser": {"Filter": {}, "Polarizer": {}},
+                "Spectrometer": {"Grating": {}, "Polarizer": {}},
+                "Detector": {"processing": {}},
+                "Spectral_image": {},
+            }
```

Comment:
Here, are all the if statement really needed? They are bound to evaluate to True since the metadata object is newly created and thus empty, if I am right.

Reply:
Originally, I didn't create a new metadata object in this method. Hence the if statements. But now they are not necessary anymore and I deleted them.